crates/ibc/src/core/ics03_connection/handler/conn_open_ack.rs

```diff
@@ -4,85 +4,129 @@ use crate::core::ics03_connection::connection::{ConnectionEnd, Counterparty, Sta
 use crate::core::ics03_connection::context::ConnectionReader;
 use crate::core::ics03_connection::error::Error;
 use crate::core::ics03_connection::events::Attributes;
-use crate::core::ics03_connection::handler::verify::{
-    check_client_consensus_height, verify_proofs,
-};
-use crate::core::ics03_connection::handler::{ConnectionIdState, ConnectionResult};
+use crate::core::ics03_connection::handler::ConnectionResult;
 use crate::core::ics03_connection::msgs::conn_open_ack::MsgConnectionOpenAck;
 use crate::events::IbcEvent;
 use crate::handler::{HandlerOutput, HandlerResult};
 use crate::prelude::*;
 
+use super::ConnectionIdState;
+
 pub(crate) fn process(
     ctx: &dyn ConnectionReader,
     msg: MsgConnectionOpenAck,
 ) -> HandlerResult<ConnectionResult, Error> {
     let mut output = HandlerOutput::builder();
 
-    // If a consensus proof is present, check that the consensus height (for
-    // client proof) in the message is not too advanced nor too old.
-    if let Some(consensus_height) = msg.consensus_height() {
-        check_client_consensus_height(ctx, consensus_height)?;
+    if msg.consensus_height > ctx.host_current_height() {
+        // Fail if the consensus height is too advanced.
+        return Err(Error::invalid_consensus_height(
+            msg.consensus_height,
+            ctx.host_current_height(),
+        ));
+    }
+
+    if msg.consensus_height < ctx.host_oldest_height() {
+        // Fail if the consensus height is too old (has been pruned).
+        return Err(Error::stale_consensus_height(
+            msg.consensus_height,
+            ctx.host_oldest_height(),
+        ));
     }
 
+    ///////////////////////////////////////////////////////////
+    // validate_self_client() verification goes here
+    // See [issue](https://github.com/cosmos/ibc-rs/issues/162)
+    ///////////////////////////////////////////////////////////
+
     // Validate the connection end.
-    let mut conn_end = ctx.connection_end(&msg.connection_id)?;
-    // A connection end must be Init or TryOpen; otherwise we return an error.
-    let state_is_consistent = conn_end.state_matches(&State::Init)
-        && conn_end.versions().contains(&msg.version)
-        || conn_end.state_matches(&State::TryOpen)
-            && conn_end.versions().get(0).eq(&Some(&msg.version));
-
-    if !state_is_consistent {
-        // Old connection end is in incorrect state, propagate the error.
+    let self_connection_end = ctx.connection_end(&msg.connection_id)?;
```

Comment:
Nit: I think I prefer `connection_end` for self connection-ends and `counterparty/expected_connection_end` for others. WDYT?

Reply:
I thought about the naming convention for a bit. The difficulty in naming here is that we have a client on a **counterparty** chain which is a light client of **self**; and vice-versa. So depending on the frame of reference chosen, e.g. `counterparty_client_state` could mean:
1. the client state of the client running on the counterparty chain
2. or the state of the "counterparty client" (that is, the client that we run of the counterparty chain) running on the host chain

So I think we should pick a frame of reference, and our naming convention should encode which one we chose. I think we should choose (1). One convention I thought of:
+ `client_state_on_{self, counterparty}`: the "on \<chain\>" makes the "chain frame of reference" clear.

So the line referred here would become:
```rust
let connection_end_on_self = ctx.connection_end(&msg.connection_id)?;
```
The most confusing line (which would have both the self and counterparty client states):
```rust
// To me, it's clear here that the light client on the host is verifying
// the client state of the light client running on the counterparty,
// and immediately makes intuitive sense
client_state_on_self.verify_client_full_state(..., client_state_on_counterparty)?;
```

WDYT? This is just the first thing I came up with, and I'm still open to other naming conventions.

Reply:
Love it. I'll implement exactly that :)

Reply:
+ `conn_open_init`: [0c6839b](https://github.com/cosmos/ibc-rs/pull/170/commits/0c6839bb87f139f912f39e9fdde6ab0e3be0df45)
+ `conn_open_try`: [858c5cf](https://github.com/cosmos/ibc-rs/pull/170/commits/858c5cfc0de103ddfb3aa743eab4e213d5dcf508) + fix [25c658d](https://github.com/cosmos/ibc-rs/pull/170/commits/25c658d6196e5f8450cd60ebc9e3d3b7117ad19a)
+ `conn_open_ack`: [73cf502](https://github.com/cosmos/ibc-rs/pull/170/commits/73cf50293a9c1383e63da4c3c935b0311ab1b214)
+ `conn_open_confirm`: [a491731](https://github.com/cosmos/ibc-rs/pull/170/commits/a49173199fb9a2ee9d3989f1766ff1ff96194660)
+ Naming convention documentation: [e549ebe](https://github.com/cosmos/ibc-rs/pull/170/commits/e549ebea336ed650ca672f19bdeb856fd570ddb0)